votingpool/input_selection.go

```diff
@@ -0,0 +1,286 @@
+		c[i].Address().branch == c[j].Address().branch &&
+		txidComparison == 0 &&
+		c[i].OutputIndex() < c[j].OutputIndex() {
+		return true
+	}
+	return false
+}
+
+// getEligibleInputs returns eligible inputs with addresses between startAddress
+// and the last used address of lastSeriesID.
+func (p *Pool) getEligibleInputs(
+	store *txstore.Store,
+	startAddress WithdrawalAddress,
+	lastSeriesID uint32,
+	dustThreshold btcutil.Amount,
+	chainHeight int32,
+	minConf int) ([]Credit, error) {
+
+	if p.GetSeries(lastSeriesID) == nil {
+		str := fmt.Sprintf("lastSeriesID (%d) does not exist", lastSeriesID)
+		return nil, newError(ErrSeriesNotExists, str, nil)
+	}
+	unspents, err := store.UnspentOutputs()
+	if err != nil {
+		return nil, newError(ErrInputSelection, "failed to get unspent outputs", err)
+	}
+	addrMap, err := groupCreditsByAddr(unspents, p.manager.ChainParams())
+	if err != nil {
+		return nil, err
+	}
+	var inputs []Credit
+	address := startAddress
+	for {
+		log.Debugf("Looking for eligible inputs at address %v", address.addrIdentifier())
+		if candidates, ok := addrMap[address.addr.EncodeAddress()]; ok {
+			var eligibles []Credit
+			for _, c := range candidates {
+				if p.isCreditEligible(c, minConf, chainHeight, dustThreshold) {
+					eligibles = append(eligibles, newCredit(c, address))
+				}
+			}
+			// Make sure the eligibles are correctly sorted.
+			sort.Sort(byAddress(eligibles))
+			inputs = append(inputs, eligibles...)
+		}
+		nAddr, err := nextAddr(p, address.seriesID, address.branch, address.index, lastSeriesID+1)
+		if err != nil {
+			return nil, newError(ErrInputSelection, "failed to get next withdrawal address", err)
+		} else if nAddr == nil {
+			log.Debugf("getEligibleInputs: reached last addr, stopping")
+			break
+		}
+		address = *nAddr
+	}
+	return inputs, nil
+}
+
+// nextAddr returns the next WithdrawalAddress according to the input selection
+// rules: http://opentransactions.org/wiki/index.php/Input_Selection_Algorithm_(voting_pools)
+// It returns nil if the new address' seriesID is >= stopSeriesID.
+func nextAddr(p *Pool, seriesID uint32, branch Branch, index Index, stopSeriesID uint32) (
+	*WithdrawalAddress, error) {
+	branch++
+	series := p.GetSeries(seriesID)
+	if series == nil {
+		return nil, newError(ErrSeriesNotExists, fmt.Sprintf("unknown seriesID: %d", seriesID), nil)
+	}
+	if int(branch) > len(series.publicKeys) {
+		highestIdx, err := p.highestUsedSeriesIndex(seriesID)
+		if err != nil {
+			return nil, err
+		}
+		if index > highestIdx {
+			seriesID++
+			log.Debugf("nextAddr(): reached last branch (%d) and highest used index (%d), "+
+				"moving on to next series (%d)", branch, index, seriesID)
+			index = 0
+		} else {
+			index++
+		}
+		branch = 0
+	}
+
+	if seriesID >= stopSeriesID {
+		return nil, nil
+	}
+
+	addr, err := p.WithdrawalAddress(seriesID, branch, index)
+	if err != nil && err.(Error).ErrorCode == ErrWithdrawFromUnusedAddr {
```

Comment:
As long as WithdrawalAddress guarantees a non-nil error will always be a `votingpool.Error` this is ok, but it will panic should any other be returned.